docs/src/developer/typesystem.md

```diff
@@ -0,0 +1,217 @@
+
+Abstract types can also belong to other abstract types, forming a tree of
+abstract types.
+
+In Nemo the most important abstract types are `Ring` and `Field`, with the
+latter belonging to the former so that all fields are rings, and the abstract
+types `RingElem` and `FieldElem` for the objects that represent elements of
+rings and fields, again with the latter abstract type belonging to the former.
+
+Because this hierarchy of abstract types must form a tree, Julia is strictly
+speaking single inheritance, as each concrete and abstract type can belong to
+at most one other abstract type. For example, one could not have a diamond of
+abstract types with `ExactField` belonging to both `Field` and `ExactRing`.
+
+### Recovering aspects of multiple inheritance in Nemo
+
+Various possibilities exist to get around the limitation that abstract types
+must form a 'tree' in Nemo and AbstractAlgebra.
+
+One such possibility is union types. If a function should accept one of a
+number of concrete or abstract types that can't all be made to belong to a
+single abstract type due to this limitation then one can use a union type.
+
+For example, Nemo defines `RingElement` to be a union of `RingElem` and all
+the Julia standard types which behave like ring elements, e.g. all `Integer`
+types and types of rationals with `Integer` components.
+
+A second feature we make use of in Nemo is parameterised types. Each concrete
+and abstract type can take one or more parameters. These parameter can be any
+other type, either concrete or abstract. For example, in Julia `Rational{T}`
+is for rationals with numerator and denominator of type `T`.
+
+A great deal of control over parameterised types is possible, e.g. one can
+restrict the type parameter `T` using a `where` clause, e.g. to write a
+function that accepts all rational types with integer components of the same
+type one can use the type `Rational{T} where T <: Integer`.
+
+Nemo makes use of such parameterised types for generic ring constructions
+such as generic polynomial rings and matrices over a given base ring. The type
+of the elements of the base ring is substituted for the parameter `T` in any
+concrete instantiation of the types `Poly{T}` and `Mat{T}`, which are defined
+in AbstractAlgebra in `src/generic/GenericTypes.jl`.
+
+The totality of all univariate polynomial types, including those of generic
+`Poly{T}` types and those coming from C libraries (such as `fmpz_poly`), is
+represented by the abstract type `PolyElem{T}` which in turn belongs to
+`RingElem`, both defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+Similarly, the totality of all matrix types, including explicit C types
+like `fmpz_mat` and the generic `Mat{T}` types is given by the abstract type
+`MatElem{T}`, again defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+This hierarchy of types allows one to write functions at any level, e.g. for
+all univariate polynomial types, just those with a given base type `T`, or
```

Comment:
We can read at https://nemocas.github.io/AbstractAlgebra.jl/latest/polynomial_rings/#Types-and-parents (emphasis mine)

> Note that the abstract types are parameterised. The type T **should usually** be the type of elements of the coefficient ring of the polynomial ring.

My impression is therefore that writing a function for polynomials with a given base type `T` is not supported, as it's not guaranteed that the first type parameter corresponds to the base ring. And "base type" is underspecified, so I think that if we document here that we can dispatch on `PolyRing{T}` for a given `T`, the quote above should be changed to 

> The type T **must** be the type of elements of the coefficient ring of the polynomial ring.

Reply:
I don't object to making that change except that 5 minutes after you make the change someone will challenge it with a counterexample. If you feel strongly about it, please open a PR to change the docs.